node.go

```diff
@@ -118,6 +119,7 @@ type nodeIterator struct {
 	nodeIteratorStep
 
 	next *BucketEntry
+	ls   ipld.LinkSystem
 }
 
 func (t *nodeIterator) Done() bool {
```

Comment:
i'm confused about the modification of iterator state happening in `Done` rather than `Next` that seems like not where i would expect that state to get changed

Reply:
Good point, happy to take a look at refactoring it.

That way we can also avoid `panic`ing and return errors instead, since `Next` function signature allows it.

Reply:
Would it really be much different if this were all done in `Next()`? It's probably not a big deal since the errors are captured, but it is slightly weird that this is all done in here.

I don't think the contract strictly says that `Done()` _must_ be called .. a naive user may just say "I know i have 3 elements, so `Next(); Next(); Next()`, which isn't going to have the imagined effect here. I think that's why it feels weird to do it in `Done()`, because it messes with the users' common mental models about how things are operating under the hood -- although this isn't a super-strong concern because we're talking about an ADL after all that's doing all sorts of messing with mental models.

:shrug: not a huge concern, just a mild smell IMO that could be deferred to later if someone feels strongly enough to deal with it.

Reply:
Resolved in https://github.com/ipld/go-ipld-adl-hamt/pull/41/commits/a427d76b436f5a52801aeca7f21a77f82557320b along with test to verify expected logic.